### ISSUE TITLE
Probably incorrect readonly modifier for pointer types with only low-level const qualifier

### ISSUE THREAD

**Issue by daiyousei-qz** [clangd/clangd#1070]
In the latest version of clangd, it seems that variables with pointer types will have its semantic token modified with readonly as long as there's one const modifier insides the type. For example, the image below shows that all declared variables are readonly.
![image](https://user-images.githubusercontent.com/12378019/159184170-87aba4f8-1763-4f1b-8002-8924e2ef9be9.png)
![image](https://user-images.githubusercontent.com/12378019/159184251-b675dd08-e54b-425b-b846-118673f5d1e1.png)

Although it might be reasonable for a `const int*` be considered readonly (I don't think so, however, as you technically can change the pointer value), this definitely doesn't work out for a type like `int const*****` IMO.

```
using ConstIntPtr      = const int*;
const int test0           = 0;
ConstIntPtr test1       = nullptr;
ConstIntPtr* test2      = nullptr;
ConstIntPtr**** test3 = nullptr;
int** const*** test4    = nullptr;
```

**System information**

Clangd version: clangd version 13.0.0

Editor/LSP plugin: VSCode 1.65.2

Operating system: Windows 11

**Comment by HighCommander4** [comment]
You're right, the current logic applies the `readonly` modifier if the type is "[const in a loose sense](https://searchfox.org/llvm/rev/10bb62319281c76827d6eda1ee70d4bac7e35935/clang-tools-extra/clangd/SemanticHighlighting.cpp#163)", with the implementation accepting a `const` after unwrapping any number of levels of pointers.

It's quite possible that this is not the most useful semantics for this modifier, though if we're going to change it we should flesh out what semantics we do want.

One possible semantics is "a pointer-like like type should only be `readonly` if a variable of that type cannot be rebound to point to a different target; the ability to modify the target is irrelevant". That would make `int * const` readonly, but not `int const *`. On the other hand, one could argue that references are conceptually pointer-like, and therefore `int &` should also be readonly by this rule because references can never be rebound.

Another possible semantics is a "a pointer-like type should be `readonly` if it cannot be rebound **or** its (direct) target cannot be modified". Then `int * const` and `int const *` would both be readonly (but not `int const **`), but arguably so should `unique_ptr<T const>`, which adds implementation challenges (we need to know about / be able to detect smart pointer types).

**Comment by daiyousei-qz** [comment]
Here's my opinion on this.

Set reference aside, I do think the ability of a name to be rebound to another object is a good matric for the `readonly` modifier. 
For example, for a variable declaration `int x = 0`, it is not `readonly` because we may rebind the name to a completely different object like 1. In the presence of pointer, say we have `int const* p1` and `int *const p2`. I think it is reasonable to regard that particular block of 4-bytes memory as an object that the name refers to. In this case, `p1` isn't read-only while `p2` is. This is also consistent with some other languages like `final` keyword in Java and `readonly` in C#.

Putting reference into consideration, a reference could be viewed as an alias of another name conceptually. In the declaration `auto& ref = x`, basically, the name `ref` should be equivalent to `x` and safely substituted with `x` if that name is accessible. 

Here's two cases:
1. `x` has type `int`, then `ref` has type `int&`.
2. `x` has type `const int`, then `ref` has type `const int&`.
Because we consider `ref` as an alias of `x`, the readonly-ness somehow should be consistent between `ref` and `x`. Therefore, I think it's reasonable to consider `const int&` being readonly while `int&` is not.

In syntax, the semantic above also shows consistency. For a particular name, whatever that could be directly used before an assign operator is not `readonly`. Otherwise, it is. Only regards to const-ness, though. A corner case could be that `T x` where `T` is a type with deleted assign operator. Should `x` be considered `readonly`?
